### DISCUSSION TITLE
setInterval and setting atom causes unnecessary rerenders

### DISCUSSION THREAD

**Discussion by nunoplima** [pmndrs/jotai#2026] [General] [answered]
In this simple counter example: https://codesandbox.io/s/react-typescript-forked-p5f947 is there a way to avoid `<Wrapper />`rerendering on every atom update? Shouldnt only` <Counter />` rerender?

**Comment by dai-shi** [comment, marked as answer]
I think it's v1 behavior which was intentional, but confused quite a few people.
That's why we developed v2 and it should work as you expect.
https://codesandbox.io/s/react-typescript-forked-33785l?file=/src/App.tsx

**Comment by nunoplima** [comment]
Hey @dai-shi, 

That's it, thanks for clarifying it, no more re-renders on the `<Wrapper />` 👍 

Related to what I'm trying to achieve, maybe you can also help me with this one, is it possible to derive from an atom that stores an object type ```{
	[id: number]: {
		isDone: boolean;
		countdownValue: string;
	};
}``` and updates every second, the `derivedAtom` would only store `{ [id: number]: boolean }`, its subscribers wouldn't rerender every second, only when one of the booleans flips.
 
Many thanks for your time and help!

**Reply by dai-shi** [reply]
It's possible and easy if you make an atom for the specific `id`.

**Reply by nunoplima** [reply]
Thanks for the quick reply!
I'm not sure I understand your suggestion 🙈, ideally I would be able to use it like ```const useDerived = () => useAtom(derivedAtom)``` and in the component ```const [derived] = useDerived()``` and ```derive[id]``` when needed.

**Reply by dai-shi** [reply]
Assuming you have `id` from props or somewhere, here's the pattern.
```js
const Component = ({ id }) => {
  const isDone = useAtomValue(
    useMemo(() => atom((get) => get(baseAtom)[id].isDone), [id])
  )
  // ...
}
```

**Reply by nunoplima** [reply]
Awesome, does make sense, thanks for that! But assuming I'm working with some sort of Table component that takes headers  as props, the headers would be something like: 
```
const headers = useMemo<ITableHeader[]>(() => {
  return [
    {
	label: 'Label Name',
        render: ({ id }) => {
	  // here we would access the id and try to get the isDone flag
	  const isDone = ...
	  return (
	    <>
	       {isDone ? <div>timer ran out</div> : <div>timer still running...</div>}
	    </>
	  )
	},
    },
    {
      ...
    }
  ]
}, [])
```

Is there a way to access the isDone from inside the useMemo avoiding multiple rerenders?

**Reply by dai-shi** [reply]
It seems like you need a bigger re-design of code structure _if_ you want jotai to optimize re-renders.
Your `render` function in headers isn't a component and you can't use hooks, and thus no ways to optimize re-renders.

fwiw, check some examples here: https://blog.axlight.com/posts/learning-react-state-manager-jotai-with-7guis-tasks/

**Reply by nunoplima** [reply]
Indeed, returning a component from the render function and using ```const isDone = useAtomValue(
    useMemo(() => atom((get) => get(baseAtom)[id].isDone), [id])
  )``` works nicely without any extra renders. Many thanks for your time and attention, keep up the great work.
